src/main/java/net/oauth/jsontoken/AsyncJsonTokenParser.java

```diff
@@ -143,4 +257,61 @@ public AsyncJsonTokenParser(
     return Futures.transformAsync(futureVerifiers, checkNullFunction, executor);
   }
 
+  /**
+   * Converts exceptions (when applicable) to {@link InvalidJsonTokenException}
+   * for better exception handling in the asynchronous parser.
+   */
+  private Exception mapException(Exception originalException) {
+    Throwable cause = originalException.getCause();
+    if (cause instanceof InvalidJsonTokenException) {
+      InvalidJsonTokenException invalidJsonTokenException = (InvalidJsonTokenException) cause;
+      if (invalidJsonTokenException.getErrorCode().equals(ErrorCode.ILLEGAL_STATE)) {
+        return new IllegalStateException(originalException);
+      }
+
+      return new InvalidJsonTokenException(
+          invalidJsonTokenException.getErrorCode(), originalException);
+    }
+
+    if (originalException instanceof SignatureException) {
+      return new InvalidJsonTokenException(ErrorCode.UNKNOWN, originalException);
+    }
+
+    if (originalException instanceof JsonParseException) {
+      return new InvalidJsonTokenException(ErrorCode.MALFORMED_TOKEN_STRING, originalException);
+    }
+
+    return originalException;
+  }
+
+  /**
+   * Rethrows {@link SignatureException}, any {@link RuntimeException}s, or any {@link Exception}
+   * where {@link Exception#getCause()} is {@link InvalidJsonTokenException}.
+   */
+  private <T> T mapExceptionAndThrow(Callable<T> callable) throws InvalidJsonTokenException {
```

Comment:
nit: Sorry for changing my previous suggestion, but now that this method's behavior has changed mapExceptionAndThrow seems a bit inaccurate because it sounds like this methods will always throw. I think both this and mapExceptionFuture could just be "mapExceptions".

Reply:
Changed and updated some documentation for them.